### PR TITLE
Makefile improvements

### PR DESCRIPTION
**Pull request by greg0ire**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -18,9 +18,10 @@ test-fix: vendor
 update-compatibility-patch:
 	@git apply tests/php-compatibility.patch
 	@echo -e "Please open your editor and apply your changes\n"
-	@until [ "$${compatibility_resolved}" == "y" ]; do read -p "Have finished your changes (y|n)? " compatibility_resolved; done && compatibility_resolved=
+	@$$EDITOR tests/expected_report.txt
```
Comment (greg0ire) [inline Makefile:21]: Can somebody that uses Mac OS test this please?

**Comment by greg0ire** [comment]
Oh wow the build is so broken… seems to have to do with Composer.

**Comment by greg0ire** [comment]
> dealerdirect/phpcodesniffer-composer-installer v0.6.2 requires composer-plugin-api ^1.0 -> found composer-plugin-api[2.0.0] but it does not match the constraint.

**Comment by greg0ire** [comment]
Related to: https://github.com/Dealerdirect/phpcodesniffer-composer-installer/issues/108

**Comment by greg0ire** [comment]
I reported the issue at https://github.com/actions/virtual-environments/issues/1000

**Comment by greg0ire** [comment]
I cherry-picked the last commit in #196 , since it is far more urgent.

File: Makefile [hunk 2]
```diff
@@ -1,6 +1,6 @@
 .PHONY: test test-report test-fix update-compatibility-patch
 
-PHP_74_OR_NEWER=`php -r "echo (int) version_compare(PHP_VERSION, '7.4', '>=');"`
+PHP_74_OR_NEWER=$(shell php -r "echo (int) version_compare(PHP_VERSION, '7.4', '>=');")
```
Comment (ostrolucky) [inline Makefile:3]: why this change?

File: Makefile (hunk 1 above)
Comment (ostrolucky) [reply Makefile:21]: What would you like me to run? just `make update-compatibility-patch`?


```❯ make update-compatibility-patch
-e Please open your editor and apply your changes

/bin/sh: tests/expected_report.txt: Permission denied
make: *** [update-compatibility-patch] Error 126
```

File: Makefile (hunk 2 above)
Comment (greg0ire) [reply Makefile:3]: Without this change, the command is run every time the variable is evaluated.

File: Makefile (hunk 1 above)
Comment (greg0ire) [reply Makefile:21]: Looks like your version of echo doesn't support `-e`. Also, looks like you don't have a `$EDITOR` variable set, can you confirm that?

Comment (ostrolucky) [reply Makefile:21]: Well yeah I confirm that. Setup should be made to deal with that though, right?

File: Makefile (hunk 2 above)
Comment (ostrolucky) [reply Makefile:3]: Interesting, is that Makefile specific thing? Because this shouldn't be the case in bash. And is same justification used for both changes: changing `` into `$()` AND adding `shell`? Where can I find something more about this?

File: Makefile (hunk 1 above)
Comment (greg0ire) [reply Makefile:21]: > Setup should be made to deal with that though, right?

Can you elaborate? I don't understand what you mean.

Comment (ostrolucky) [reply Makefile:21]: I mean Makefile should handle the case when $EDITOR is not defined

File: Makefile (hunk 2 above)
Comment (greg0ire) [reply Makefile:3]: Yes, [it's very much Makefile specific](https://www.gnu.org/software/make/manual/html_node/Flavors.html#Flavors). It looks like an alternative would be to use `:=` note that [using backticks is deprecated in favor of `$()` in bash](https://unix.stackexchange.com/questions/126927/have-backticks-i-e-cmd-in-sh-shells-been-deprecated).
Also note that `$(shell foo)` is a function call and does not result in make calling an hypothetical `shell` binary. I think I am going to go with `:=`, that should be less complicated.

Comment (greg0ire) [reply Makefile:3]: Ok I'm going to go with both: https://stackoverflow.com/questions/1435861/computing-makefile-variable-on-assignment

File: Makefile [hunk 3]
```diff
@@ -18,9 +18,10 @@ test-fix: vendor
...
 	@git apply tests/php-compatibility.patch
 	@echo -e "Please open your editor and apply your changes\n"
-	@until [ "$${compatibility_resolved}" == "y" ]; do read -p "Have finished your changes (y|n)? " compatibility_resolved; done && compatibility_resolved=
+	@$$EDITOR tests/expected_report.txt
 	@git diff -- tests/expected_report.txt tests/fixed > .tmp-patch && mv .tmp-patch tests/php-compatibility.patch && git apply -R tests/php-compatibility.patch
 	@git commit -m 'Update compatibility patch' tests/php-compatibility.patch
 
 vendor: composer.json
 	composer update
+	touch vendor
```
Comment (ostrolucky) [inline Makefile:27]: What about this, why are we assuming composer update won't create vendor folder?

Comment (greg0ire) [reply Makefile:27]: We are not assuming it won't create the vendor folder, but we are assuming (righfully) that it will not change its modification date if it already exists.

File: Makefile (hunk 1 above)
Comment (greg0ire) [reply Makefile:21]: Ah is see, with `EDITOR ?= vim` for instance?

Comment (greg0ire) [reply Makefile:21]: Seems a bit forceful, I will just drop that commit.

**Review by ostrolucky — DISMISSED** [review]

File: Makefile (hunk 3 above)
Comment (ostrolucky) [reply Makefile:27]: So this updates timestamp of vendor folder, which will make sure dependant targets like `test-report` will run even when vendor wasn't changed 👍

Comment (morozov) [reply Makefile:27]: If we only want to modify the date, should we use `touch -c` instead? Additionally, are we interested in the directory itself or in all its files (something like `$(shell find vendor -type f)`)? That way, we wouldn't have to touch the directory.

Comment (greg0ire) [reply Makefile:27]: If we only want to modify the date, should we use touch -c instead?

@ostrolucky or you @morozov if you use Mac OS can you test `touch -c` or better, `touch --no-create` with `sh` to see if it is portable? I have no idea.

>  Additionally, are we interested in the directory itself or in all its files (something like $(shell find vendor -type f))? That way, we wouldn't have to touch the directory.

What are you suggesting here? `make` only checks the `vendor` directory's timestamp (because it is the target). It will not check any of the underlying files.

Comment (morozov) [reply Makefile:27]: > @morozov if you use Mac OS

That sounds insulting. No.

> What are you suggesting here?

I meant it to be used in the dependency context e.g. `target: foo bar $(shell  find vendor -type f))`. But I'm not really sure how it's used here, so nevermind.

**Review by SenseException — DISMISSED** [review]

File: Makefile (hunk 3 above)
Comment (greg0ire) [reply Makefile:27]: The contents of the `vendor` directory only depends on 2 files `composer.json` and `composer.lock`. And by depends, `make` means they are read (by Composer in this case) when generating that directory.

Comment (greg0ire) [reply Makefile]: @ostrolucky ran a quick test `--no-create` is not supported, but `-c` is, let's go with that.

**Review by ostrolucky — APPROVED** [review]

**Review by SenseException — APPROVED** [review]

**Review by morozov — APPROVED** [review]

**Review by beberlei — APPROVED** [review]

**Review by lcobucci — APPROVED** [review]

**Comment by greg0ire** [comment]
@lcobucci why did you assign this to yourself? Do you mean to test it locally before merging, maybe?

**Comment by lcobucci** [comment]
I was going to test it locally and merge it but then life happened.

**Comment by lcobucci** [comment]
@greg0ire sorry, mate! :ship:
